subsys/net/lib/wireguard/wg.c

```diff
@@ -0,0 +1,2078 @@
+{
+	NET_PKT_DATA_ACCESS_DEFINE(wg_access, struct wg_msg_hdr);
+	enum net_verdict verdict = NET_DROP;
+	struct sockaddr_storage my_addr_s = { 0 };
+	struct sockaddr *my_addr = (struct sockaddr *)&my_addr_s;
+	struct sockaddr_storage addr_s = { 0 };
+	struct sockaddr *addr = (struct sockaddr *)&addr_s;
+	union net_proto_header *udp_hdr;
+	union net_ip_header *ip_hdr;
+	struct wg_msg_hdr *hdr;
+	struct wg_peer *peer;
+	size_t len, hdr_len;
+	int ret;
+
+	if (pkt->buffer == NULL) {
+		goto drop;
+	}
+
+	len = net_pkt_get_len(pkt);
+
+	ip_hdr = net_pkt_wg_ip_hdr(pkt);
+	udp_hdr = net_pkt_wg_udp_hdr(pkt);
+
+	if (net_pkt_family(pkt) == AF_INET) {
+		if (len < NET_IPV4UDPH_LEN + sizeof(struct wg_msg_hdr)) {
+			NET_DBG("DROP: Too short Wireguard header");
+			goto drop;
+		}
+
+		net_pkt_cursor_init(pkt);
+
+		hdr_len = net_pkt_ip_hdr_len(pkt) + NET_UDPH_LEN;
+
+		if (net_pkt_skip(pkt, hdr_len)) {
+			NET_DBG("DROP: Too short %s packet", "IPv4");
+			goto drop;
+		}
+
+		memcpy(&net_sin(addr)->sin_addr, &ip_hdr->ipv4->src,
+		       sizeof(struct in_addr));
+		net_sin(addr)->sin_port = udp_hdr->udp->src_port;
+		addr->sa_family = AF_INET;
+
+		memcpy(&net_sin(my_addr)->sin_addr, &ip_hdr->ipv4->dst,
+		       sizeof(struct in_addr));
+		net_sin(my_addr)->sin_port = udp_hdr->udp->dst_port;
+		net_sin(my_addr)->sin_family = AF_INET;
+
+	} else if (net_pkt_family(pkt) == AF_INET6) {
+		if (len < NET_IPV6UDPH_LEN + sizeof(struct wg_msg_hdr)) {
+			NET_DBG("DROP: Too short Wireguard header");
+			goto drop;
+		}
+
+		net_pkt_cursor_init(pkt);
+
+		hdr_len = net_pkt_ip_hdr_len(pkt) + NET_UDPH_LEN;
+
+		if (net_pkt_skip(pkt, hdr_len)) {
+			NET_DBG("DROP: Too short %s packet", "IPv6");
+			goto drop;
+		}
+
+		memcpy(&net_sin6(addr)->sin6_addr, &ip_hdr->ipv6->src,
+		       sizeof(struct in6_addr));
+		net_sin6(addr)->sin6_port = udp_hdr->udp->src_port;
+		addr->sa_family = AF_INET6;
+
+		memcpy(&net_sin6(my_addr)->sin6_addr, &ip_hdr->ipv6->dst,
+		       sizeof(struct in6_addr));
+		net_sin6(my_addr)->sin6_port = udp_hdr->udp->dst_port;
+		net_sin6(my_addr)->sin6_family = AF_INET6;
+	} else {
+		return -EAFNOSUPPORT;
+	}
+
+	hdr = (struct wg_msg_hdr *)net_pkt_get_data(pkt, &wg_access);
+	if (!hdr) {
+		NET_DBG("DROP: NULL Wireguard header");
+		goto drop;
+	}
+
+	if (!(hdr->reserved[0] == 0 && hdr->reserved[1] == 0 &&
+	      hdr->reserved[2] == 0)) {
+		NET_DBG("DROP: Invalid Wireguard header");
+		goto drop;
+	}
+
+	/* At this point we don't know if this came from a valid peer */
+	peer = peer_lookup_by_virtual_iface(iface);
+	if (peer == NULL) {
+		NET_DBG("DROP: Peer not found for interface %d",
+			net_if_get_by_iface(iface));
+		goto drop;
+	}
+
+	if (peer->ctx == NULL) {
+		NET_DBG("Invalid configuration");
+		goto drop;
+	}
+
+	if (hdr->type == MESSAGE_HANDSHAKE_INITIATION) {
+		if (!IS_ENABLED(CONFIG_WIREGUARD_SERVER)) {
+			NET_DBG("DROP: %s not supported in client mode",
+				"handshake init");
+			goto drop;
+		}
```

Comment:
is the protocol happy with this option? trying to run it with `CONFIG_WIREGUARD_SERVER=n`, seems like the peer really likes to init on their side sometimes, even after the connection has been established I occasionally get some

```
[00:03:55.173,000] <dbg> wireguard: wg_ctrl_recv: (rx_q[0]): DROP: handshake init not supported in client mode
```

and no traffic flows anymore, peer sees an increasing `latest handshake` in `wg` and that's it

Reply:
Yeah something's off, with this enabled and some constant traffic from the peer the connection is stable, as soon as that stops data stops flowing after a while until the peer re-initiates

Reply:
There is some issue when zephyr tries to connect to peer, I am investigating it.

Reply:
> trying to run it with `CONFIG_WIREGUARD_SERVER=n`

That is a good question if we should have that server option. My motivation for it was to restrict the incoming traffic if the zephyr device is only suppose to connect out. And one way to support that was not to allow incoming init if server is not enabled, not sure if there is better alternative for that, or if we really need the server option at all.